gems/sorbet-runtime/lib/types/private/final.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+# typed: false
+
+module T::Private::Final
+  def self.declare(mod)
+    if !mod.is_a?(Module)
+      raise "#{mod.name} is not a module or method and cannot be declared as final"
+    end
+    if final_module?(mod)
+      raise "#{mod.name} was already declared as final and cannot be re-declared as final"
+    end
+    if T::AbstractUtils.abstract_module?(mod)
+      raise "#{mod.name} was already declared as abstract and cannot be declared as final"
+    end
+    if mod.is_a?(Class)
+      orig_inherited = T::Private::ClassUtils.replace_method(mod.singleton_class, :inherited) do |arg|
+        orig_inherited.bind(self).call(arg)
+        raise "#{mod.name} was declared as final and cannot be inherited from"
+      end
+    end
+    mark_as_final_module(mod)
+    mark_as_final_module(mod.singleton_class)
+    T::Private::Methods.install_hooks(mod)
+  end
+
+  def self.final_module?(mod)
+    mod.instance_variable_defined?(:@sorbet_final_module) # rubocop:disable PrisonGuard/NoLurkyInstanceVariableAccess
```

Comment:
You can delete the PrisonGuard bits now that we're not running in pay-server. We can probably delete all of the existing ones too, but that's separate.